common/lib/host_list_provider/monitoring/cluster_topology_monitor.ts

```diff
@@ -0,0 +1,482 @@
+  private readonly _pluginService: PluginService;
+  private readonly _hostListProvider: MonitoringRdsHostListProvider;
+  private readonly refreshRateMs: number;
+  private readonly highRefreshRateMs: number;
+
+  private topologyMap: CacheMap<string, HostInfo[]>;
+  private writerHostInfo: HostInfo = null;
+  private isVerifiedWriterConnection: boolean = false;
+  private monitoringClient: ClientWrapper = null;
+  private highRefreshRateEndTime: number = 0;
+  private highRefreshPeriodAfterPanicMs: number = 30000; // 30 seconds.
+  private ignoreNewTopologyRequestsEndTime: number = 0;
+  private ignoreTopologyRequestMs: number = 10000; // 10 seconds.
+
+  // Controls for stopping the ClusterTopologyMonitor run.
+  private stopMonitoring: boolean = false;
+  private readonly runPromise: Promise<void>;
+
+  // Tracking of the host monitors.
+  private hostMonitors: Map<string, HostMonitor> = new Map();
+  public hostMonitorsWriterClient = null;
+  public hostMonitorsWriterInfo: HostInfo = null;
+  public hostMonitorsReaderClient = null;
+  public hostMonitorsLatestTopology: HostInfo[] = null;
+  // Controls for stopping all the host monitors run.
+  public hostMonitorsStop: boolean = false;
+  private untrackedPromises: Promise<void>[] = [];
+
+  // Signals to other methods that asynchronous operations have completed/should be completed.
+  private requestToUpdateTopology: boolean = false;
+  private releaseTopologyUpdate: () => void;
+  private topologyUpdated = new Promise<void>((done) => {
+    this.releaseTopologyUpdate = () => {
+      done();
+    };
+  });
+
+  constructor(
+    clusterId: string,
+    topologyMap: CacheMap<string, HostInfo[]>,
+    initialHostSpec: HostInfo,
+    props,
+    pluginService: PluginService,
+    hostListProvider: MonitoringRdsHostListProvider,
+    refreshRateMs: number,
+    highRefreshRateMs: number
+  ) {
+    this.clusterId = clusterId;
+    this.topologyMap = topologyMap;
+    this.initialHostInfo = initialHostSpec;
+    this._pluginService = pluginService;
+    this._hostListProvider = hostListProvider;
+    this._properties = props;
+    this.refreshRateMs = refreshRateMs;
+    this.highRefreshRateMs = highRefreshRateMs;
+    this.runPromise = this.run();
+  }
+
+  get hostListProvider(): MonitoringRdsHostListProvider {
+    return this._hostListProvider;
+  }
+  get pluginService(): PluginService {
+    return this._pluginService;
+  }
+  get properties(): Map<string, any> {
+    return this._properties;
+  }
+
+  async close(): Promise<void> {
+    this.stopMonitoring = true;
+    this.hostMonitorsStop = true;
+    await Promise.all(this.runPromise ? this.untrackedPromises.concat(this.runPromise) : this.untrackedPromises);
+    this.untrackedPromises = [];
+    this.hostMonitors.clear();
+  }
+
+  async forceMonitoringRefresh(shouldVerifyWriter: boolean, timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts = this.topologyMap.get(this.clusterId);
+    if (currentHosts) {
```

Comment:
Since currentHosts is either `HostInfo[]` or `null` lets strictly check for null here.

```suggestion
    if (currentHosts !== null) {
```